pvlib/temperature.py

```diff
@@ -423,3 +424,206 @@ def faiman(poa_global, temp_air, wind_speed=1.0, u0=25.0, u1=6.84):
+    ----------
+    .. [1] Fuentes, M. K. A Simplifed Thermal Model for Flat-Plate
+           Photovoltaic Arrays. SAND85-0330. 1987.
+           http://prod.sandia.gov/techlib/access-control.cgi/1985/850330.pdf
+    .. [2] Dobos, A. P. PVWatts Version 5 Manual. NREL/TP-6A20-62641. 2014.
+           doi:10.2172/1158421.
+    """
+    # ported from the FORTRAN77 code provided in Appendix A of Fuentes 1987;
+    # nearly all variable names are kept the same for ease of comparison.
+
+    boltz = 5.669e-8
+    emiss = emissivity
+    absorp = absorption
+    xlen = hydraulic_diameter
+    cap0 = 11000
+    tinoct = inoct + 273.15
+
+    # convective coefficient of top surface of module at NOCT
+    windmod = 1.0
+    tave = (tinoct + 293.15) / 2
+    hconv = _fuentes_hconv(tave, windmod, tinoct, tinoct - 293.15, xlen, False)
+
+    # determine the ground temperature ratio and the ratio of the total
+    # convection to the top side convection
+    hground = emiss * boltz * (tinoct**2 + 293.15**2) * (tinoct + 293.15)
+    backrat = (
+        absorp * 800.0
+        - emiss * boltz * (tinoct**4 - 282.21**4)
+        - hconv * (tinoct - 293.15)
+    ) / ((hground + hconv) * (tinoct - 293.15))
+    tground = (tinoct**4 - backrat * (tinoct**4 - 293.15**4))**0.25
+    tground = np.clip(tground, 293.15, tinoct)
+
+    tgrat = (tground - 293.15) / (tinoct - 293.15)
+    convrat = (absorp * 800 - emiss * boltz * (
+        2 * tinoct**4 - 282.21**4 - tground**4)) / (hconv * (tinoct - 293.15))
+
+    # adjust the capacitance (thermal mass) of the module based on the INOCT.
+    # It is a function of INOCT because high INOCT implies thermal coupling
+    # with the racking (e.g. roofmount), so the thermal mass is increased.
+    # `cap` has units J/(m^2 C) -- see Table 3, Equations 26 & 27
+    cap = cap0
+    if tinoct > 321.15:
+        cap = cap * (1 + (tinoct - 321.15) / 12)
+
+    # iterate through timeseries inputs
+    sun0 = 0
+    tmod0 = 293.15
+
+    # n.b. the way Fuentes calculates the first timedelta makes it seem like
+    # the value doesn't matter -- rather than recreate it here, just assume
+    # it's the same as the second timedelta:
+    timedelta_hours = np.diff(poa_global.index).astype(float) / 1e9 / 60 / 60
+    timedelta_hours = np.append([timedelta_hours[0]], timedelta_hours)
+
+    df = pd.DataFrame({
+        'tamb': temp_air + 273.15,
+        'sun': poa_global * absorp,
+        'wind_speed': wind_speed,
+        'dtime': timedelta_hours,
+    })
+
+    # Two of the calculations are easily vectorized, so precalculate them:
+    # sky temperature -- Equation 24
+    df['tsky'] = 0.68 * (0.0552 * df['tamb']**1.5) + 0.32 * df['tamb']
+    # wind speed at module height -- Equation 22
+    # not sure why the 1e-4 factor is included -- maybe the equations don't
+    # behave well if wind == 0?
+    df['windmod'] = df['wind_speed'] * (module_height/wind_height)**0.2 + 1e-4
+
+    tmod0 = 293.15
+
+    for idx, row in df.iterrows():
```

Comment:
Not a recommendation, but a question: consider `scipy.integrate.solve_ivp` here? The RHS of the diff. eq. in Fuentes is easily vectorized, might a performance gain over the loops.

Reply:
Interesting idea.  I'm a little rusty on ODE problems but... isn't the approach taken here not only doing IVP integration (outer loop) but also fixed point iteration (inner loop)?  Is it possible to do both with `solve_ivp`?  If so it would be a good alternative to this manual looping business, but it would take me some time to figure out.

Reply:
I think `scipy.integrate.solve_ivp` _could_ be used here but not without taking apart and reworking some of the supporting calculations. The two-loop approach in Fuentes is a customized solver for the ODE in his paper, taking advantage of an approximation that linearizes the equation. I don't see any conceptual reason why `solve_ivp` with defaults couldn't solve the equation. But, OK to use the Fuentes technique unless performance becomes unsatisfactory.